### DISCUSSION TITLE
How to Learn SOFA for Real-Time Deformable Shape Sensing as a Beginner?

### DISCUSSION THREAD

**Discussion by sodasoda520** [sofa-framework/sofa#4955] [Build / Config / Environment] [answered]
Hello SOFA community,

I am a complete beginner with SOFA and I'm interested in achieving a rather complex task. My goal is to develop a real-time deformable shape sensing system for a soft body. Specifically, I want to:

1. Track the displacement of specific marked points on a soft sheet in real-time（ using camrea）.
2. Update the soft body's shape in the SOFA simulation according to these tracked displacements.
3. Export the updated positions of all FEM model nodes and the forces acting on them.

Given my beginner status, I would appreciate guidance on the best learning path to achieve this. Could you please suggest:

- Essential SOFA tutorials or documentation that I should start with.
- The key SOFA components and modules I should understand.
- Any advice on integrating external tracking data into SOFA.
- Tips for implementing real-time updates to the FEM model.

Thank you in advance for your help and for pointing me in the right direction!

Best regards

**Comment by bakpaul** [comment, marked as answer]
Hello ! 

You project is close to what I worked on during my phd. There are many ways to do that, it depends on what you want to achieve :
1) is it a forward simulation where you just want to impose the measured displacement ot the object and use the simulation to extrapolate the shape
2) Is it an inverse simulation where you are trying to find the displacement of some boundary condition to make the FE model match the tracking data

Depending on that you might take different paths and use different plugins. 

For both you'll need to understand a bit the mechanism of mapping, because your tracked points might not lie on points of the mesh and thus you'll need to map them on the deformable mesh to use them. 

For the first situation : the core will be enough for you. I suggest that you look at the examples using constraints, either weak constraints (springs), projective constraints or Lagrangian constraints. Because this is how you'll apply your measured data to the object. 

For the second one, you'll nee to use an inverse framework. This is available in the [SoftRobots.Inverse](https://github.com/SofaDefrost/SoftRobots.Inverse) plugin. 

Then for the tutorials, you can find some of them on our youtube channel, especially our [introduction course](https://www.youtube.com/watch?v=GKH1lTvO_YM). We will make multiple training session during the year, I can keep you updated if you want for the next one (I think that it will be on Soft robotics).

**Reply by sodasoda520** [reply]
Thank you so much for your detailed and insightful response. Your explanation has given me a much clearer understanding of the different paths I could take.

My research direction may indeed involve soft robotics in the future, but for now, I’m focusing on implementing a soft shape servo demo using SOFA. Specifically, as you mentioned, this would involve forward simulation. My plan is to build a finite element model of the soft body, use a camera to capture the displacements of surface markers after deformation, then feed these displacements into the SOFA model. The model would then update, and I would export the positions of all nodes for further calculations.

However, I still have some uncertainties:

Real-Time Data Handling: Does SOFA support the real-time transmission and reception of data, such as continuously updating the model with the displacement data and exporting the updated node positions?

Environment Configuration:

Is it better to use Windows or Linux for this project?
Should I download the latest version of SOFA, and would you recommend using a pre-compiled version or compiling it from source?
What programming language would be most suitable for this task? I have some experience with Python and C++, but I’m not sure which would be best for working with SOFA.
Are there any specific plugins that I need to install to achieve this functionality?
Thank you again for your guidance, and I really appreciate any further advice you can provide.

Best regards

**Reply by damienmarchal** [reply]
Hi, 

Regarding this point: 
> Real-Time Data Handling: Does SOFA support the real-time transmission and reception of data, such as continuously updating the model with the displacement data and exporting the updated node positions?

There is no problem in having sofa exchanging live data to/from third party application/devices.

The easiest and most flexible approach, which does not requires a pulgin, is to use a [Python Controller](https://sofapython3.readthedocs.io/en/latest/content/modules/Sofa/generated/Sofa.Core/classes/Sofa.Core.Controller.html). A Controller in SOFA is a special component you can add in scenes to implement custom behavior at specific events of the simulation (eg: begin/end animation, begin/end rendering, fixed interval, on key pressed/release, mouse click, ... ). So it can be used to exchange the data between the sofa's data field and the in/out data streams/usb/serial port/network at every simulation step.

At [DEFROST](https://www.defrost.inria.fr/) we often starts with python and  [ZeroMQ](https://zeromq.org/) or ROS as communication library. Controllers can also be implemented in C++ (if python is not an option). 

Using windows or linux will not makes a big difference as long as there is there is a library for the communication protocol you planned to use.  

DM. 





 

> 
> Environment Configuration:
> 
> Is it better to use Windows or Linux for this project? Should I download the latest version of SOFA, and would you recommend using a pre-compiled version or compiling it from source? What programming language would be most suitable for this task? I have some experience with Python and C++, but I’m not sure which would be best for working with SOFA. Are there any specific plugins that I need to install to achieve this functionality? Thank you again for your guidance, and I really appreciate any further advice you can provide.
> 
> Best regards

**Reply by sodasoda520** [reply]
Thank you for your guidance—it's been incredibly helpful. I’m looking forward to diving into this project with the confidence that I’m on the right track.
